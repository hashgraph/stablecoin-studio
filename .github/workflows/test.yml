name: test
on:
  pull_request:
    branches: [ main, release/**, feature/** ]
  push:
    branches: [ main, release/** ]
    tags: [ v* ]
jobs:
  test-node:
    name: Unit Tests
    runs-on: ubuntu-latest
    steps:
      - name: Use Node.js TLS 16.x
        uses: actions/setup-node@v1
        with:
          node-version: 16.x
      - name: Checkout repository
        uses: actions/checkout@v2
      - name: Install dependencies
        run: npm i
      - name: Build
        run: npm run build
      - name: Test
        run: npm run test