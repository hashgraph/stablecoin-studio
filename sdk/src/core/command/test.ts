/* eslint-disable @typescript-eslint/no-empty-interface */

interface CommandResponse {}
interface ICommand {}

export class Command<T> implements ICommand {
	private $resultType!: T;
}

export interface ICommandBus {
	execute<X>(command: Command<X>): Promise<X>;
}

export type IInferringCommandHandler<CommandType extends Command<unknown>> =
	CommandType extends Command<infer ResultType>
		? {
				execute(command: CommandType): Promise<ResultType>;
		  }
		: never;

class ConcreteCommandResponse implements CommandResponse {
	constructor(public readonly status: string) {}
}

class ConcreteCommand extends Command<ConcreteCommandResponse> {
	constructor(public readonly value: string) {
		super();
	}
}

class CommandHandler<CommandType extends Command<unknown>>
	implements IInferringCommandHandler<ConcreteCommand>
{
	execute(command: ConcreteCommand): Promise<ConcreteCommandResponse> {
		throw new Error('Method not implemented.');
	}
}

class CommandBus implements ICommandBus {
	execute<X>(command: Command<X>): Promise<X> {
		throw new Error('Method not implemented.');
	}
}

// Test

const bus = new CommandBus();
const res = bus.execute(new ConcreteCommand('1'));
