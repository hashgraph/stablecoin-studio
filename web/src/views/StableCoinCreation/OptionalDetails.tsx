import { Heading, Stack, VStack } from '@chakra-ui/react';
import type { Control, FieldValues } from 'react-hook-form';
import { useWatch } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import InputNumberController from '../../components/Form/InputNumberController';
import { SelectController } from '../../components/Form/SelectController';

interface OptionalDetailsProps {
	control: Control<FieldValues>;
}

const OptionalDetails = (props: OptionalDetailsProps) => {
	const { control } = props;
	const { t } = useTranslation(['global', 'stableCoinCreation']);

	const supplyTypes = [
		{
			value: 0,
			label: t('stableCoinCreation:optionalDetails.infinite'),
		},
		{
			value: 1,
			label: t('stableCoinCreation:optionalDetails.finite'),
		},
	];

	const selectorStyle = {
		wrapper: {
			border: '1px',
			borderColor: 'brand.black',
			borderRadius: '8px',
			height: 'min',
		},
		menuList: {
			maxH: '220px',
			overflowY: 'auto',
			bg: 'brand.white',
			boxShadow: 'down-black',
			p: 2,
			zIndex: 99,
		},
		valueSelected: {
			fontSize: '14px',
			fontWeight: '500',
		},
	};

	const isSupplyTypeFinite =
		useWatch({
			control,
			name: 'supplyType',
		})?.value === 1;

	const initialSupply = useWatch({
		control,
		name: 'initialSupply',
	});

	return (
		<VStack h='full' justify={'space-between'} pt='80px'>
			<Stack minW={400}>
				<Heading
					data-testid='title'
					fontSize='16px'
					fontWeight='600'
					mb={10}
					lineHeight='15.2px'
					textAlign={'left'}
				>
					{t('stableCoinCreation:optionalDetails.title')}
				</Heading>
				<Stack as='form' spacing={6}>
					<InputNumberController
						rules={{
							required: t(`global:validations.required`),
						}}
						isRequired
						control={control}
						name={'initialSupply'}
						label={t('stableCoinCreation:optionalDetails.initialSupply')}
						placeholder={t('stableCoinCreation:optionalDetails.placeholder', {
							placeholder: t('stableCoinCreation:optionalDetails.initialSupply'),
						})}
					/>
					<SelectController
						control={control}
						name={'supplyType'}
						options={supplyTypes}
						label={t('stableCoinCreation:optionalDetails.typeSupply')}
						placeholder={t('stableCoinCreation:optionalDetails.typeSupplyPlaceholder')}
						overrideStyles={selectorStyle}
						addonLeft={true}
						variant='unstyled'
						defaultValue={'0'}
					/>
					{isSupplyTypeFinite && (
						<InputNumberController
							rules={{
								required: t(`global:validations.required`),
								validate: {
									quantityOverTotalSupply: (value: number) => {
										return (
											(initialSupply && initialSupply <= value) ||
											t('global:validations.overMaxSupply')
										);
									},
								},
							}}
							isRequired
							control={control}
							name={'maxSupply'}
							label={t('stableCoinCreation:optionalDetails.maxSupply')}
							placeholder={t('stableCoinCreation:optionalDetails.placeholder', {
								placeholder: t('stableCoinCreation:optionalDetails.maxSupply'),
							})}
						/>
					)}
					<InputNumberController
						rules={{
							required: t(`global:validations.required`),
						}}
						isRequired
						control={control}
						name={'decimals'}
						label={t('stableCoinCreation:optionalDetails.decimals')}
						placeholder={t('stableCoinCreation:optionalDetails.placeholder', {
							placeholder: t('stableCoinCreation:optionalDetails.decimals'),
						})}
						maxValue={18}
						initialValue={6}
					/>
				</Stack>
			</Stack>
		</VStack>
	);
};

export default OptionalDetails;
