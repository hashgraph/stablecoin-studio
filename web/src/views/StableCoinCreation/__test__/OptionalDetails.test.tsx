import { FieldValues, useForm } from 'react-hook-form';
import { render } from '../../../test/index';
import translations from '../../../translations/en/stableCoinCreation.json';
import OptionalDetails from '../OptionalDetails';
import { CreateRequest } from '@hashgraph-dev/stablecoin-npm-sdk';

jest.mock('react-hook-form', () => ({
	...jest.requireActual('react-hook-form'),
	Controller: () => <></>,
	useForm: () => ({
		getValues: () => ({
			maxSupply: 1000,
			initialSupply: 1000,
			supplyType: 1,
		}),
	}),
}));

const form = useForm<FieldValues>({
	mode: 'onChange',
	defaultValues: {
		initialSupply: 1000,
	},
});

const { control } = form;
const request = new CreateRequest({
	name: 'name',
	symbol: 'symbol',
	decimals: 6,
	createReserve: false,
});

describe(`<${OptionalDetails.name} />`, () => {
	test('should render correctly', () => {
		const component = render(<OptionalDetails control={control} request={request} form={form} />);

		expect(component.asFragment()).toMatchSnapshot();
	});

	test('should has title', () => {
		const component = render(<OptionalDetails control={control} request={request} form={form} />);

		const header = component.getByTestId('title');
		expect(header).toHaveTextContent(translations.optionalDetails.title);
	});
});
