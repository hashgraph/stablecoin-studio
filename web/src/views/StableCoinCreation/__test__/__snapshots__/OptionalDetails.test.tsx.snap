// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OptionalDetails /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-1wbej9n"
  >
    <div
      class="chakra-stack css-8csoc2"
    >
      <h2
        class="chakra-heading css-p7jgmy"
        data-testid="title"
      >
        Configure Supply & Metadata
      </h2>
      <form
        class="chakra-stack css-ahthbn"
      />
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
